x-pack/legacy/plugins/apm/public/new-platform/plugin.tsx

```diff
@@ -68,5 +69,8 @@ export class Plugin {
       </KibanaCoreContextProvider>,
       document.getElementById(REACT_APP_ROOT_ID)
     );
+
+    // create static index pattern and store as saved object. Not needed by APM UI but for legacy reasons in Discover, Dashboard etc.
+    createStaticIndexPattern(core.http);
```

Comment:
The static index pattern is created on every page refresh but the app doesn't wait for the response, so I think it's fine.

Reply:
Should this check whether there is data?

Reply:
Do you see any advantages of making an explicit check? The reason I didn't do this is to avoid the additional roundtrip. Currently the index pattern is created with `overwrite: false` which has the same effect afaict.

Reply:
What I understood from our conversation is that other plugins will encounter issues if we have a static index pattern without any data being available (which is the reason why we cannot do this on plugin startup). Is the same concern not applicable here? Do we assume that anyone that opens the APM plugin has APM data?

Reply:
Ah, now I get what you mean! 

tldr: I think we should yes! 👍 

Below I wrote the background for the original decision and why we shouldn't check for data but I changed my mind halfway. I still think the background might be useful to have so leaving it as optional reading ;)

--

I tend to agree with you - but it's a trade-off, and I'm not sure it's worth it.
As you mentioned the we tried creating the index pattern on plugin startup but this change was [reverted](https://github.com/elastic/kibana/issues/37499) because it caused to issues:

**1. Discover would show an error if the index pattern existed without there being matching indices**
![](https://user-images.githubusercontent.com/1238659/58897334-e8382000-86ac-11e9-8ec9-53a673ad18f4.png)

**2. Kibana Welcome Screen was disabled**
The Kibana welcome screen would only be displayed if no index existed. Since the apm index pattern was created on startup at least one index pattern would always exist.
https://github.com/elastic/kibana/issues/37498

Both these issues affected all users regardless if they actively used APM because APM UI is enabled by default.
By creating the index pattern lazily when a user enters the UI we've made this scenario much less severe. It can still happen (and probably will) but it will only affect users who actively clicks on the APM UI tab. And it won't prevent the Kibana Home screen from showing up on the very first visit.

My reason for not wanting to check for data is that we'd make the check on every (full) page load of APM UI, and I'm wondering how costly that'd be. However, as I'm writing this I realize that with `terminate_after: 1` the query should be quite performant, so that shouldn't be a blocker.

Reply:
Thanks for the background! 👍